src/main/java/io/numaproj/numaflow/sink/SinkService.java

```diff
@@ -5,59 +5,107 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
 import static io.numaproj.numaflow.function.v1.UserDefinedFunctionGrpc.getMapFnMethod;
 
 class SinkService extends UserDefinedSinkGrpc.UserDefinedSinkImplBase {
-    private final SinkHandler sinkHandler;
+    private static final Logger logger = Logger.getLogger(SinkService.class.getName());
+    // it will never be smaller than one
+    private final ExecutorService sinkTaskExecutor = Executors
+            .newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
+    private final long SHUTDOWN_TIME = 30;
+    private SinkHandler sinkHandler;
+
+    public SinkService() {
+    }
 
-    public SinkService(SinkHandler sinkHandler) {
+    public void setSinkHandler(SinkHandler sinkHandler) {
         this.sinkHandler = sinkHandler;
     }
 
     /**
-     * Applies a function to each datum element.
+     * Applies a function to each datum element in the stream.
      */
-    public void sinkFn(
-            Udsink.DatumList request,
-            StreamObserver<Udsink.ResponseList> responseObserver) {
+    @Override
+    public StreamObserver<Udsink.Datum> sinkFn(StreamObserver<Udsink.ResponseList> responseObserver) {
         if (this.sinkHandler == null) {
-            io.grpc.stub.ServerCalls.asyncUnimplementedUnaryCall(
+            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(
                     getMapFnMethod(),
                     responseObserver);
-            return;
         }
+        SinkDatumStreamImpl sinkDatumStream = new SinkDatumStreamImpl();
 
-        // get DatumList from request
-        var numElem = request.getElementsCount();
-        var handlerDatumList = new Udsink.Datum[numElem];
-        for (int i = 0; i < numElem; i++) {
-            var d = request.getElements(i);
-            handlerDatumList[i] = Udsink.Datum.newBuilder()
-                    .setId(d.getId())
-                    .setValue(d.getValue())
-                    .setEventTime(d.getEventTime())
-                    .setWatermark(d.getWatermark())
-                    .build();
-        }
+        Future<List<Response>> result = sinkTaskExecutor.submit(() -> sinkHandler.HandleDo(
+                sinkDatumStream));
 
-        // process DatumList
-        Response[] responses = sinkHandler.HandleDo(handlerDatumList);
+        return new StreamObserver<Udsink.Datum>() {
+            @Override
+            public void onNext(Udsink.Datum d) {
+                // get Datum from request
+                Udsink.Datum handlerDatum = Udsink.Datum.newBuilder()
+                        .setId(d.getId())
+                        .setValue(d.getValue())
+                        .setEventTime(d.getEventTime())
+                        .setWatermark(d.getWatermark())
```

Comment:
wondering if we should remove the watermark and not expose that to the udf.

Reply:
we expose it in other SDKs